File: tests/excel_report.py
import os
import openpyxl
from openpyxl.styles import Font, Alignment, PatternFill, Border, Side
from datetime import datetime

class ExcelReport:
    def __init__(self):
        self.wb = openpyxl.Workbook()
        self.excel_dir = os.path.join(os.path.dirname(os.path.abspath(__file__)), 'excel')
        os.makedirs(self.excel_dir, exist_ok=True)
        
        # Remove default sheet
        default_sheet = self.wb.active
        self.wb.remove(default_sheet)
        
        self.tests = []
        self.logs = []
        self.env_details = {"browser": "Chrome", "mode": "Headless", "url": "http://localhost:8000"}

    def set_env_details(self, browser, mode, url):
        self.env_details = {
            "browser": browser,
            "mode": mode,
            "url": url
        }

    def add_test(self, test_id, module, scenario, status, duration, start_time, end_time, failure_reason="", stack_trace="", screenshot_path=""):
        self.tests.append({
            "id": test_id,
            "module": module,
            "scenario": scenario,
            "status": status,
            "duration": duration, # in seconds
            "start_time": start_time,
            "end_time": end_time,
            "failure_reason": failure_reason,
            "stack_trace": stack_trace,
            "screenshot_path": screenshot_path
        })

    def add_log(self, timestamp, test_id, step, result, remarks=""):
        self.logs.append({
            "timestamp": timestamp,
            "test_id": test_id,
            "step": step,
            "result": result,
            "remarks": remarks
        })

    def generate_report(self):
        # Create worksheets
        summary_sheet = self.wb.create_sheet(title='Summary')
        test_cases_sheet = self.wb.create_sheet(title='Test Cases')
        failed_tests_sheet = self.wb.create_sheet(title='Failed Tests')
        execution_logs_sheet = self.wb.create_sheet(title='Execution Logs')
        
        # Styles
        header_font = Font(name='Segoe UI', size=11, bold=True, color='FFFFFF')
        header_fill = PatternFill(start_color='1F4E78', end_color='1F4E78', fill_type='solid') # Steel blue
        header_alignment = Alignment(vertical='center', horizontal='center', wrap_text=True)
        
        thin_border_side = Side(style='thin', color='E0E0E0')
        border_style = Border(top=thin_border_side, bottom=thin_border_side, left=thin_border_side, right=thin_border_side)
        
        # 1. Summary Sheet
        summary_sheet.views.sheetView[0].showGridLines = True
        summary_sheet.append([
            'Execution Date', 'Browser / Driver', 'Execution Mode', 'Target URL', 
            'Total Tests', 'Passed', 'Failed', 'Pass Percentage', 'Execution Duration'
        ])
        
        total = len(self.tests)
        passed = sum(1 for t in self.tests if t['status'].lower() == 'passed')
        failed = sum(1 for t in self.tests if t['status'].lower() == 'failed')
        pass_percentage = f"{round((passed / total) * 100)}%" if total > 0 else "0%"
        total_duration = sum(t['duration'] for t in self.tests)
        duration_formatted = f"{total_duration:.2f}s"
        
        summary_sheet.append([
            datetime.now().strftime("%Y-%m-%d %H:%M:%S"),
            self.env_details['browser'],
            self.env_details['mode'],
            self.env_details['url'],
            total,
            passed,
            failed,
            pass_percentage,
            duration_formatted
        ])
        
        # Style Summary
        summary_sheet.row_dimensions[1].height = 28
        summary_sheet.row_dimensions[2].height = 24
        for col_idx in range(1, 10):
            cell = summary_sheet.cell(row=1, column=col_idx)
            cell.font = header_font
            cell.fill = header_fill
            cell.alignment = header_alignment
            
            val_cell = summary_sheet.cell(row=2, column=col_idx)
            val_cell.font = Font(name='Segoe UI', size=10)
            val_cell.border = border_style
            val_cell.alignment = Alignment(horizontal='center', vertical='center')
            
            if col_idx == 8: # Pass Percentage coloring
                val_cell.font = Font(name='Segoe UI', size=10, bold=True, color='2E7D32' if passed == total and total > 0 else 'C62828')
                
        # 2. Test Cases Sheet
        test_cases_sheet.views.sheetView[0].showGridLines = True
        test_cases_sheet.append([
            'Test ID', 'Module', 'Scenario', 'Status', 'Start Time', 'End Time', 'Duration'
        ])
        test_cases_sheet.row_dimensions[1].height = 28
        
        for col_idx in range(1, 8):
            cell = test_cases_sheet.cell(row=1, column=col_idx)
            cell.font = header_font
            cell.fill = header_fill
            cell.alignment = header_alignment
            
        for idx, t in enumerate(self.tests, start=2):
            test_cases_sheet.append([
                t['id'],
                t['module'],
                t['scenario'],
                t['status'],
                t['start_time'].strftime("%Y-%m-%d %H:%M:%S") if t['start_time'] else "",
                t['end_time'].strftime("%Y-%m-%d %H:%M:%S") if t['end_time'] else "",
                f"{t['duration']:.2f}s"
            ])
            test_cases_sheet.row_dimensions[idx].height = 22
            for col_idx in range(1, 8):
                cell = test_cases_sheet.cell(row=idx, column=col_idx)
                cell.font = Font(name='Segoe UI', size=10)
                cell.border = border_style
                cell.alignment = Alignment(vertical='center')
                if col_idx in [1, 4, 7]:
                    cell.alignment = Alignment(horizontal='center', vertical='center')
                if col_idx == 4:
                    if t['status'].lower() == 'passed':
                        cell.font = Font(name='Segoe UI', bold=True, color='2E7D32')
                    elif t['status'].lower() == 'failed':
                        cell.font = Font(name='Segoe UI', bold=True, color='C62828')
                    else:
                        cell.font = Font(name='Segoe UI', bold=True, color='F9A825')

        # 3. Failed Tests Sheet
        failed_tests_sheet.views.sheetView[0].showGridLines = True
        failed_tests_sheet.append([
            'Test ID', 'Scenario', 'Failure Reason', 'Screenshot Path', 'Failure Details / Stack Trace'
        ])
        failed_tests_sheet.row_dimensions[1].height = 28
        
        failed_fill = PatternFill(start_color='C62828', end_color='C62828', fill_type='solid') # Red header
        for col_idx in range(1, 6):
            cell = failed_tests_sheet.cell(row=1, column=col_idx)
            cell.font = header_font
            cell.fill = failed_fill
            cell.alignment = header_alignment
            
        failed_list = [t for t in self.tests if t['status'].lower() == 'failed']
        for idx, t in enumerate(failed_list, start=2):
            failed_tests_sheet.append([
                t['id'],
                t['scenario'],
                t['failure_reason'],
                t['screenshot_path'],
                t['stack_trace']
            ])
            failed_tests_sheet.row_dimensions[idx].height = 26
            for col_idx in range(1, 6):
                cell = failed_tests_sheet.cell(row=idx, column=col_idx)
                cell.font = Font(name='Segoe UI', size=9)
                cell.border = border_style
                cell.alignment = Alignment(vertical='center', wrap_text=True)
                if col_idx == 1:
                    cell.alignment = Alignment(horizontal='center', vertical='center')

        # 4. Execution Logs Sheet
        execution_logs_sheet.views.sheetView[0].showGridLines = True
        execution_logs_sheet.append([
            'Timestamp', 'Test ID', 'Step Description', 'Result', 'Remarks'
        ])
        execution_logs_sheet.row_dimensions[1].height = 28
        for col_idx in range(1, 6):
            cell = execution_logs_sheet.cell(row=1, column=col_idx)
            cell.font = header_font
            cell.fill = header_fill
            cell.alignment = header_alignment
            
        for idx, l in enumerate(self.logs, start=2):
            execution_logs_sheet.append([
                l['timestamp'],
                l['test_id'],
                l['step'],
                l['result'],
                l['remarks']
            ])
            execution_logs_sheet.row_dimensions[idx].height = 20
            for col_idx in range(1, 6):
                cell = execution_logs_sheet.cell(row=idx, column=col_idx)
                cell.font = Font(name='Segoe UI', size=9)
                cell.border = border_style
                cell.alignment = Alignment(vertical='center')
                if col_idx in [1, 2, 4]:
                    cell.alignment = Alignment(horizontal='center', vertical='center')
                if col_idx == 4:
                    res = l['result'].lower()
                    if 'pass' in res or 'success' in res:
                        cell.font = Font(name='Segoe UI', size=9, bold=True, color='2E7D32')
                    elif 'fail' in res or 'error' in res:
                        cell.font = Font(name='Segoe UI', size=9, bold=True, color='C62828')
                        
        # Auto-adjust column widths for all sheets
        for sheet in [summary_sheet, test_cases_sheet, failed_tests_sheet, execution_logs_sheet]:
            for col in sheet.columns:
                max_len = 0
                col_letter = openpyxl.utils.get_column_letter(col[0].column)
                for cell in col:
                    if cell.value is not None:
                        # Handle multiple lines in cell for length estimation
                        lines = str(cell.value).split('\n')
                        for line in lines:
                            max_len = max(max_len, len(line))
                sheet.column_dimensions[col_letter].width = max(max_len + 3, 12)
                
        # Write file
        file_path = os.path.join(self.excel_dir, 'Web_E2E_Report.xlsx')
        self.wb.save(file_path)
        return file_path

excel_report = ExcelReport()

File: mobile-e2e/utilities/excelReport.js
import ExcelJS from 'exceljs';
import path from 'path';
import fs from 'fs';
import logger from './logger.js';

export class ExcelReport {
  constructor() {
    this.workbook = new ExcelJS.Workbook();
    this.excelDir = 'excel';
    
    // Ensure output folder exists
    if (!fs.existsSync(this.excelDir)) {
      fs.mkdirSync(this.excelDir, { recursive: true });
    }

    this.summarySheet = null;
    this.testCasesSheet = null;
    this.failedTestsSheet = null;
    this.executionLogsSheet = null;

    this.tests = [];
    this.logs = [];
    this.deviceDetails = { name: 'Android Device', version: '11.0' };
  }

  setDeviceDetails(name, version) {
    this.deviceDetails = { name, version };
  }

  addTest(test) {
    this.tests.push({
      id: test.id || `TC_${this.tests.length + 1}`,
      module: test.module || 'General',
      scenario: test.scenario || 'E2E Scenario',
      device: this.deviceDetails.name,
      version: this.deviceDetails.version,
      status: test.status || 'Passed',
      startTime: test.startTime || new Date(),
      endTime: test.endTime || new Date(),
      duration: test.duration || 0,
      failureReason: test.failureReason || '',
      screenshotPath: test.screenshotPath || '',
      activityName: test.activityName || '',
      stackTrace: test.stackTrace || ''
    });
  }

  addLog(timestamp, testName, step, result, remarks = '') {
    this.logs.push({
      timestamp: timestamp || new Date().toISOString(),
      testName: testName || 'Global Setup',
      step,
      result,
      remarks
    });
  }

  async generateReport() {
    logger.info('Compiling test results and generating ExcelJS Workbook...');

    // Initialize workbook worksheets
    this.summarySheet = this.workbook.addWorksheet('Summary');
    this.testCasesSheet = this.workbook.addWorksheet('Test Cases');
    this.failedTestsSheet = this.workbook.addWorksheet('Failed Tests');
    this.executionLogsSheet = this.workbook.addWorksheet('Execution Logs');

    // 1. Setup Styles
    const headerStyle = {
      font: { name: 'Segoe UI', size: 11, bold: true, color: { argb: 'FFFFFFFF' } },
      fill: { type: 'pattern', pattern: 'solid', fgColor: { argb: 'FF1F4E78' } }, // Steel blue
      alignment: { vertical: 'middle', horizontal: 'center' },
      border: {
        top: { style: 'thin', color: { argb: 'FFD3D3D3' } },
        bottom: { style: 'medium', color: { argb: 'FF1F4E78' } }
      }
    };

    const borderStyle = {
      top: { style: 'thin', color: { argb: 'FFE0E0E0' } },
      bottom: { style: 'thin', color: { argb: 'FFE0E0E0' } },
      left: { style: 'thin', color: { argb: 'FFE0E0E0' } },
      right: { style: 'thin', color: { argb: 'FFE0E0E0' } }
    };

    // 2. Populate Sheet 1 - Summary
    this.summarySheet.columns = [
      { header: 'Metric', key: 'metric', width: 25 },
      { header: 'Value', key: 'value', width: 18 },
      { header: 'Percentage', key: 'percentage', width: 18 }
    ];

    const total = this.tests.length;
    const passed = this.tests.filter(t => t.status.toLowerCase() === 'passed').length;
    const failed = this.tests.filter(t => t.status.toLowerCase() === 'failed').length;
    const skipped = this.tests.filter(t => t.status.toLowerCase() === 'skipped').length;
    const passedPercentDetailed = total > 0 ? `${((passed / total) * 100).toFixed(2)}%` : '0.00%';
    const failedPercentDetailed = total > 0 ? `${((failed / total) * 100).toFixed(2)}%` : '0.00%';
    const skippedPercentDetailed = total > 0 ? `${((skipped / total) * 100).toFixed(2)}%` : '0.00%';
    const totalDurationMs = this.tests.reduce((acc, t) => acc + t.duration, 0);
    const avgDurationMs = total > 0 ? (totalDurationMs / total).toFixed(2) : '0.00';

    this.summarySheet.addRow({ metric: 'Total Tests', value: total, percentage: '100%' });
    this.summarySheet.addRow({ metric: 'Passed', value: passed, percentage: passedPercentDetailed });
    this.summarySheet.addRow({ metric: 'Failed', value: failed, percentage: failedPercentDetailed });
    this.summarySheet.addRow({ metric: 'Skipped', value: skipped, percentage: skippedPercentDetailed });
    this.summarySheet.addRow({ metric: 'Success Rate', value: passedPercentDetailed, percentage: '' });
    this.summarySheet.addRow({ metric: 'Total Duration (ms)', value: totalDurationMs, percentage: '' });
    this.summarySheet.addRow({ metric: 'Average Test Duration', value: parseFloat(avgDurationMs), percentage: '' });

    // Style Summary Headers and Cells
    this.summarySheet.getRow(1).eachCell((cell) => { Object.assign(cell, headerStyle); });
    this.summarySheet.getRow(1).height = 28;

    // Row 2: Total Tests (Green)
    this.summarySheet.getRow(2).eachCell((cell) => {
      cell.font = { name: 'Segoe UI', size: 10, bold: true, color: { argb: 'FF2E7D32' } };
      cell.border = borderStyle;
      cell.alignment = { horizontal: 'center', vertical: 'middle' };
      if (cell.col === 1) cell.alignment = { horizontal: 'left', vertical: 'middle' };
    });
    this.summarySheet.getRow(2).height = 22;

    // Row 3: Passed (Red)
    this.summarySheet.getRow(3).eachCell((cell) => {
      cell.font = { name: 'Segoe UI', size: 10, bold: true, color: { argb: 'FFC62828' } };
      cell.border = borderStyle;
      cell.alignment = { horizontal: 'center', vertical: 'middle' };
      if (cell.col === 1) cell.alignment = { horizontal: 'left', vertical: 'middle' };
    });
    this.summarySheet.getRow(3).height = 22;

    // Row 4: Failed (Yellow)
    this.summarySheet.getRow(4).eachCell((cell) => {
      cell.font = { name: 'Segoe UI', size: 10, bold: true, color: { argb: 'FFF9A825' } };
      cell.border = borderStyle;
      cell.alignment = { horizontal: 'center', vertical: 'middle' };
      if (cell.col === 1) cell.alignment = { horizontal: 'left', vertical: 'middle' };
    });
    this.summarySheet.getRow(4).height = 22;

    // Rows 5-8: normal styling
    for (let r = 5; r <= 8; r++) {
      this.summarySheet.getRow(r).eachCell((cell) => {
        cell.font = { name: 'Segoe UI', size: 10 };
        cell.border = borderStyle;
        cell.alignment = { horizontal: 'center', vertical: 'middle' };
        if (cell.col === 1) cell.alignment = { horizontal: 'left', vertical: 'middle' };
      });
      this.summarySheet.getRow(r).height = 22;
    }

    // 3. Populate Sheet 2 - Test Cases
    this.testCasesSheet.columns = [
      { header: 'Test ID', key: 'id', width: 12 },
      { header: 'Module', key: 'module', width: 16 },
      { header: 'Scenario', key: 'scenario', width: 32 },
      { header: 'Device', key: 'device', width: 16 },
      { header: 'Status', key: 'status', width: 12 },
      { header: 'Start Time', key: 'startTime', width: 22 },
      { header: 'End Time', key: 'endTime', width: 22 },
      { header: 'Duration', key: 'duration', width: 14 }
    ];

    this.testCasesSheet.getRow(1).eachCell((cell) => { Object.assign(cell, headerStyle); });
    this.testCasesSheet.getRow(1).height = 28;

    this.tests.forEach((t) => {
      const row = this.testCasesSheet.addRow({
        id: t.id,
        module: t.module,
        scenario: t.scenario,
        device: t.device,
        status: t.status,
        startTime: t.startTime.toLocaleString(),
        endTime: t.endTime.toLocaleString(),
        duration: `${(t.duration / 1000).toFixed(2)}s`
      });
      row.height = 22;
      row.eachCell((cell) => {
        cell.font = { name: 'Segoe UI', size: 10 };
        cell.border = borderStyle;
        cell.alignment = { vertical: 'middle' };
        if (cell.col === 1 || cell.col === 4 || cell.col === 5 || cell.col === 8) {
          cell.alignment = { horizontal: 'center', vertical: 'middle' };
        }
        // Status formatting (col 5 now)
        if (cell.col === 5) {
          if (t.status === 'Passed') {
            cell.font = { name: 'Segoe UI', bold: true, color: { argb: 'FF2E7D32' } }; // Green
          } else if (t.status === 'Failed') {
            cell.font = { name: 'Segoe UI', bold: true, color: { argb: 'FFC62828' } }; // Red
          } else {
            cell.font = { name: 'Segoe UI', bold: true, color: { argb: 'FFF9A825' } }; // Yellow/Orange
          }
        }
      });
    });

    // 4. Populate Sheet 3 - Failed Tests
    this.failedTestsSheet.columns = [
      { header: 'Test Name', key: 'testName', width: 30 },
      { header: 'Failure Reason', key: 'reason', width: 35 },
      { header: 'Screenshot Path', key: 'screenshot', width: 40 },
      { header: 'Device', key: 'device', width: 16 },
      { header: 'Android Version', key: 'version', width: 15 },
      { header: 'Activity Name', key: 'activity', width: 25 },
      { header: 'Execution Time', key: 'execTime', width: 22 },
      { header: 'Failure Details', key: 'details', width: 45 }
    ];

    this.failedTestsSheet.getRow(1).eachCell((cell) => {
      // Dark red header for failed tests
      Object.assign(cell, headerStyle);
      cell.fill = { type: 'pattern', pattern: 'solid', fgColor: { argb: 'FFC62828' } };
    });
    this.failedTestsSheet.getRow(1).height = 28;

    const failedList = this.tests.filter(t => t.status === 'Failed');
    failedList.forEach((t) => {
      const row = this.failedTestsSheet.addRow({
        testName: t.scenario,
        reason: t.failureReason,
        screenshot: t.screenshotPath,
        device: t.device,
        version: t.version,
        activity: t.activityName,
        execTime: t.startTime.toLocaleString(),
        details: t.stackTrace
      });
      row.height = 26;
      row.eachCell((cell) => {
        cell.font = { name: 'Segoe UI', size: 9 };
        cell.border = borderStyle;
        cell.alignment = { vertical: 'middle', wrapText: true };
        if (cell.col === 4 || cell.col === 5 || cell.col === 7) {
          cell.alignment = { horizontal: 'center', vertical: 'middle' };
        }
      });
    });

    // 5. Populate Sheet 4 - Execution Logs
    this.executionLogsSheet.columns = [
      { header: 'Timestamp', key: 'timestamp', width: 22 },
      { header: 'Test Name', key: 'testName', width: 30 },
      { header: 'Step', key: 'step', width: 45 },
      { header: 'Result', key: 'result', width: 14 },
      { header: 'Remarks', key: 'remarks', width: 35 }
    ];

    this.executionLogsSheet.getRow(1).eachCell((cell) => { Object.assign(cell, headerStyle); });
    this.executionLogsSheet.getRow(1).height = 28;

    this.logs.forEach((l) => {
      const row = this.executionLogsSheet.addRow({
        timestamp: l.timestamp,
        testName: l.testName,
        step: l.step,
        result: l.result,
        remarks: l.remarks
      });
      row.height = 20;
      row.eachCell((cell) => {
        cell.font = { name: 'Segoe UI', size: 9 };
        cell.border = borderStyle;
        cell.alignment = { vertical: 'middle' };
        if (cell.col === 1 || cell.col === 4) {
          cell.alignment = { horizontal: 'center', vertical: 'middle' };
        }
        if (cell.col === 4) {
          if (l.result.toLowerCase().includes('pass') || l.result.toLowerCase() === 'success') {
            cell.font = { name: 'Segoe UI', size: 9, bold: true, color: { argb: 'FF2E7D32' } };
          } else if (l.result.toLowerCase().includes('fail') || l.result.toLowerCase() === 'error') {
            cell.font = { name: 'Segoe UI', size: 9, bold: true, color: { argb: 'FFC62828' } };
          }
        }
      });
    });

    // Write file
    const filePath = path.join(this.excelDir, 'Mobile_E2E_Report.xlsx');
    await this.workbook.xlsx.writeFile(filePath);
    logger.info(`Excel report successfully generated at: ${filePath}`);
    return filePath;
  }
}

// Export singleton instance for easy imports across tests
export const excelReport = new ExcelReport();
export default excelReport;
